creator-node/src/services/stateMachineManager/monitoring/StateMonitoringQueue.js

```diff
@@ -0,0 +1,254 @@
+  }
+
+  logWarn(msg) {
+    logger.warn(`StateMonitoringQueue WARNING: ${msg}`)
+  }
+
+  logError(msg) {
+    logger.error(`StateMonitoringQueue ERROR: ${msg}`)
+  }
+
+  makeQueue(redisHost, redisPort) {
+    // Settings config from https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#advanced-settings
+    return new BullQueue(STATE_MONITORING_QUEUE_NAME, {
+      redis: {
+        host: redisHost,
+        port: redisPort
+      },
+      defaultJobOptions: {
+        removeOnComplete: QUEUE_HISTORY,
+        removeOnFail: QUEUE_HISTORY
+      },
+      settings: {
+        // Should be sufficiently larger than expected job runtime
+        lockDuration: STATE_MONITORING_QUEUE_MAX_JOB_RUNTIME_MS,
+        // We never want to re-process stalled jobs
+        maxStalledCount: 0
+      }
+    })
+  }
+
+  registerQueueEventHandlers(queue) {
+    // Add handlers for logging
+    queue.on('global:waiting', (jobId) => {
+      this.log(`Queue Job Waiting - ID ${jobId}`)
+    })
+    queue.on('global:active', (jobId, jobPromise) => {
+      this.log(`Queue Job Active - ID ${jobId}`)
+    })
+    queue.on('global:lock-extension-failed', (jobId, err) => {
+      this.logError(
+        `Queue Job Lock Extension Failed - ID ${jobId} - Error ${err}`
+      )
+    })
+    queue.on('global:stalled', (jobId) => {
+      this.logError(`stateMachineQueue Job Stalled - ID ${jobId}`)
+    })
+    queue.on('global:error', (error) => {
+      this.logError(`Queue Job Error - ${error}. Queuing another job...`)
+    })
+
+    // Re-queue a new job when the current job fails or succeeds
+    queue.on('completed', (job, result) => {
+      this.log(
+        `Queue Job Completed - ID ${job?.id} - Result ${JSON.stringify(
+          result
+        )}. Queuing another job...`
+      )
+      if (result?.jobFailed) {
+        this.enqueueJobAfterSuccess(queue, job, result)
+      } else {
+        this.enqueueJobAfterFailure(queue, job)
+      }
+    })
+    queue.on('failed', (job, err) => {
+      this.logError(
+        `Queue Job Failed - ID ${job?.id} - Error ${err}. Queuing another job...`
+      )
+      this.enqueueJobAfterFailure(queue, job)
+    })
+  }
+
+  /**
+   * Enqueues a job that processes the next slice of users after the slice
+   * that the previous job sucessfully processed.
+   * @param queue the StateMonitoringQueue to enqueue a job
+   * @param successfulJobData the jobData of the previous job that succeeded
+   * @param successfulJobResult the result of the previous job that succeeded
+   */
+  enqueueJobAfterSuccess(queue, successfulJobData, successfulJobResult) {
+    const {
+      data: { discoveryNodeEndpoint, moduloBase, currentModuloSlice }
+    } = successfulJobData
+    const { lastProcessedUserId } = successfulJobResult
+
+    queue.add({
+      startTime: Date.now(),
+      lastProcessedUserId,
+      discoveryNodeEndpoint,
+      moduloBase,
+      currentModuloSlice: (currentModuloSlice + 1) % moduloBase
```

Comment:
do we still need modulo logic anywhere in this? we're not using it anymore right

Reply:
i see you have a linear card to clean this up, but since this is a separate queue and old logic is still there, its prob fine to remove it just to make the new code more streamlined

Reply:
synced offline. keeping for now